cmd/agent/api/grpc.go

```diff
@@ -208,43 +217,80 @@ func (s *serverSecure) GetConfigs(ctx context.Context, in *pb.GetConfigsRequest)
 	if s.configService == nil {
 		log.Debug("Remote configuration service not initialized")
 		return errors.New("remote config service not initialized")
 	}
 
-	ctx := out.Context()
+	ctx := channel.Context()
 	configs := make(chan *pb.ConfigResponse, 1)
+	reqs := make(chan *pb.SubscribeConfigRequest, 1)
+
+	go s.listenReqs(ctx, channel, reqs)
+	go s.sendConfigs(ctx, channel, configs)
 
-	log.Debugf("New remote configuration subscriber request for product %s", in.Product)
-	subscriber := remoteconfig.NewSubscriber(in.Product, time.Second, func(config *pb.ConfigResponse) error {
-		log.Debug("Pushing configuration for gRPC client")
+	for {
 		select {
-		case configs <- config:
-			log.Debug("Pushed configuration to gRPC client")
+		case req := <-reqs:
+			if !s.configService.HasSubscriber(req.Product) {
+				log.Debugf("New remote configuration subscriber request for product %s", req.Product)
+				subscriber := remoteconfig.NewChanSubscriber(req.Product, time.Second, configs)
+				log.Debugf("New remote configuration subscriber for product %s", req.Product)
+				s.configService.RegisterSubscriber(subscriber)
+				defer s.configService.UnregisterSubscriber(subscriber)
+			}
+		case <-ctx.Done():
+			log.Info("Unsubscribing gRPC client")
 			return nil
-		default:
-			return errors.New("failed to notify gRPC subscriber")
 		}
-	})
-
-	log.Debugf("New remote configuration subscriber for product %s", in.Product)
-	s.configService.RegisterSubscriber(subscriber)
-	defer s.configService.UnregisterSubscriber(subscriber)
+	}
+}
 
+func (s *serverSecure) sendConfigs(ctx context.Context, channel pb.AgentSecure_GetConfigUpdatesServer, configs chan *pb.ConfigResponse) {
 	for {
 		log.Debug("Streaming config to gRPC client")
 		select {
 		case config := <-configs:
-			log.Debugf("Sending configuration for product %s", in.Product)
-			if err := out.Send(config); err != nil {
+			log.Debug("Sending configuration")
+			if err := channel.Send(config); err != nil {
 				log.Errorf("error sending config event: %s", err)
-				return nil
+				time.Sleep(retrySleepDuration)
+				continue
 			}
 		case <-ctx.Done():
-			log.Infof("Unsubscribing gRPC client for product %s", in.Product)
-			return nil
+			log.Info("Unsubscribing gRPC client")
+			return
+		}
+	}
+}
+
+func (s *serverSecure) listenReqs(ctx context.Context, channel pb.AgentSecure_GetConfigUpdatesServer, reqs chan *pb.SubscribeConfigRequest) {
+	log.Debug("Starting to listen for subscribe config requests")
+	for {
+		req, err := channel.Recv()
+		if err == io.EOF {
+			continue
+		} else if err != nil {
+			log.Errorf("Error receiving config request: %s", err)
+		}
+
+		log.Debug("Adding subscribe config request")
+		reqs <- req
+
+		if req != nil && req.TracerInfo != nil {
+			err := s.configService.TracerInfos.AddTracer(req.TracerInfo)
+			if err != nil {
+				log.Errorf("Error sending tracer info: %s", err)
+			}
```

Comment:
Should be a debug log - we don't want to spam the logs when we fill up the cache.

Reply:
Should we track the tracer info here or in `GetConfigUpdates`? I think this could just forward messages and we can have the request handling logic in the other method - don't feel strongly about it though 🙂

Reply:
These are actually two different gRPC methods: https://github.com/DataDog/datadog-agent/blob/main/pkg/proto/datadog/api/v1/api.proto#L95-L108
`GetConfigs` at the moment won't have a tracer, but I still wanted to add it here for future use cases. But I agree that it shouldn't spam the logs.

Reply:
I can't see the diff you're talking about anymore :( Let me know if I answered a different question than what you actually asked ;)